beacon_node/beacon_chain/src/fork_choice.rs

```diff
@@ -0,0 +1,196 @@
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    MissingBlock(Hash256),
+    MissingState(Hash256),
+    BackendError(String),
+    BeaconStateError(BeaconStateError),
+    StoreError(StoreError),
+}
+
+pub struct ForkChoice<T: BeaconChainTypes> {
+    backend: T::LmdGhost,
+    /// Used for resolving the `0x00..00` alias back to genesis.
+    ///
+    /// Does not necessarily need to be the _actual_ genesis, it suffices to be the finalized root
+    /// whenever the struct was instantiated.
+    genesis_block_root: Hash256,
+}
+
+impl<T: BeaconChainTypes> ForkChoice<T> {
+    /// Instantiate a new fork chooser.
+    ///
+    /// "Genesis" does not necessarily need to be the absolute genesis, it can be some finalized
+    /// block.
+    pub fn new(
+        store: Arc<T::Store>,
+        genesis_block: &BeaconBlock,
+        genesis_block_root: Hash256,
+    ) -> Self {
+        Self {
+            backend: T::LmdGhost::new(store, genesis_block, genesis_block_root),
+            genesis_block_root,
+        }
+    }
+
+    pub fn find_head(&self, chain: &BeaconChain<T>) -> Result<Hash256> {
+        let start_slot = |epoch: Epoch| epoch.start_slot(T::EthSpec::slots_per_epoch());
+
+        // From the specification:
+        //
+        // Let justified_head be the descendant of finalized_head with the highest epoch that has
+        // been justified for at least 1 epoch ... If no such descendant exists,
+        // set justified_head to finalized_head.
+        let (start_state, start_block_root, start_block_slot) = {
+            let state = chain.current_state();
+
+            let (block_root, block_slot) =
+                if state.current_epoch() + 1 > state.current_justified_epoch {
+                    (
+                        state.current_justified_root,
+                        start_slot(state.current_justified_epoch),
+                    )
+                } else {
+                    (state.finalized_root, start_slot(state.finalized_epoch))
+                };
+
+            let block = chain
+                .store
+                .get::<BeaconBlock>(&block_root)?
+                .ok_or_else(|| Error::MissingBlock(block_root))?;
+
+            // Resolve the `0x00.. 00` alias back to genesis
+            let block_root = if block_root == Hash256::zero() {
+                self.genesis_block_root
+            } else {
+                block_root
+            };
+
+            let state = chain
+                .store
+                .get::<BeaconState<T::EthSpec>>(&block.state_root)?
+                .ok_or_else(|| Error::MissingState(block.state_root))?;
+
+            (state, block_root, block_slot)
+        };
+
+        // A function that returns the weight for some validator index.
+        let weight = |validator_index: usize| -> Option<u64> {
+            start_state
+                .validator_registry
+                .get(validator_index)
+                .and_then(|v| Some(v.effective_balance))
```

Comment:
`.and_then(|..| Some(..))` could be written `.map(|..| ..)`, doesn't matter too much though